Detect class inited status more correctly in prestub/jitinterface

davidwrighton opened the pull request:
Fix #104247 by providing a new helper function which means what the old `IsClassInited` helper meant.

This should not have a semantic effect on anything, but the performance should be a bit better.

davidwrighton committed:
Fix #104247 by providing a new helper function which means what the old IsClassInited helper meant

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -3795,6 +3795,24 @@ void MethodTable::EnsureStaticDataAllocated()
+    if (IsClassInited())
+        return TRUE;
+    
+    AttemptToPreinit();
```
`AttemptToPreinit` is a bit complex method. We may end up calling it here repeatedly number of times.

Would it better to set the initialized flag during `SetIsStaticDataAllocated` so that we do not need to be repeating the preinit check and all calls to `IsClassInited()` will take care of the pre-initialization automatically?

```
    inline void SetIsStaticDataAllocated()
    {
        LIMITED_METHOD_CONTRACT;
        DWORD dwFlags = enum_flag_IsStaticDataAllocated;
        if (IsPreinited())
            dwFlags |= enum_flag_Initialized;
        InterlockedOr((LONG*)&m_dwFlags, (LONG)dwFlags);
    }
```

davidwrighton replied:
Mmm... seems like a decent idea, I'll look into that once I understand the arm64 failure seen in the tests.

jkotas replied:
The linux arm64 failure is happening on multiple PRs. Opened https://github.com/dotnet/runtime/issues/104263.

davidwrighton committed:
Tweak so that we compute the Init state when we ensure that the static data is allocated
This allows the IsClassInitedOrPreinited function to be implemented with a single memory read in all repeated use cases

davidwrighton committed:
Fix parentheses issue

davidwrighton committed:
- Fix issue where the gc statics field for strings isn't being checked correctly
- AND a very subtle race condition where we need to set the IsClassInited flag bit on dynamic statics BEFORE setting the MethodTable level one

EgorBo commented:
@MihuBot

EgorBo commented:
@MihuBot -nocctors
